File: Python Blind 75/Arrays/Two Sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # Time: O(n), Space: O(n) 
        hashMap = {} # val, index

        for i, val in enumerate(nums): # enumerate allows looping through the index and value of each element
            diff = target - val
            if diff in hashMap:
                return [hashMap[diff], i]
            hashMap[val] = i

File: Java Blind 75/Arrays/Two Sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        // Time O(n):
        // Because it is only a one pass through the array

        // Space O(1):
        // Because the amount of variables stays the same no 
        // matter the length of the inputs

        // key is the number at index and value is the index itself
        HashMap<Integer, Integer> hash = new HashMap<>();
        
        for (int i = 0; i < nums.length; i++) {
            if (hash.containsKey(target - nums[i])) {
                return new int[]{hash.get(target - nums[i]), i};
            }
            hash.put(nums[i], i);
        }
        return new int[]{};
    }
}